{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56281b13-d0f1-40ff-abc0-e6de240494c3",
   "metadata": {},
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "\n",
    "Ans :- Inheritance and Polymorphism: Inheriting from the Exception class allows your custom exception to inherit the properties and behaviors of the base class. This includes methods such as __str__, which controls the string representation of the exception when it is printed, and attributes like args, which stores the arguments passed to the exception constructor. By using the Exception class as the base, your custom exception can take advantage of these standard features, ensuring consistent behavior with other exceptions in the Python ecosystem.\n",
    "\n",
    "Compatibility and Consistency: By using the Exception class, your custom exception adheres to the established conventions and patterns of exception handling in Python. It ensures compatibility with existing code and libraries that expect exceptions to be derived from the Exception class. Following this convention makes your custom exception easier to understand and integrate with other codebases, as developers are familiar with the standard exception hierarchy.\n",
    "\n",
    "Exception Handling: The Exception class provides a common interface for exception handling. By inheriting from this class, your custom exception can be caught using generic except statements that handle exceptions at a higher level. This simplifies exception handling, allowing you to catch and handle multiple types of exceptions using a single except block if necessary.\n",
    "\n",
    "Code Readability and Maintainability: Using the Exception class as the base for custom exceptions enhances code readability and maintainability. Developers who encounter your custom exception will immediately recognize it as an exception and understand its purpose. It also helps in self-documenting code, making it easier for others (including your future self) to understand and work with your code.\n",
    "\n",
    "Integration with Libraries and Frameworks: Many Python libraries and frameworks are designed to work with the Exception class and its subclasses. By using the Exception class as the base for your custom exception, you ensure that it seamlessly integrates with existing error handling mechanisms and frameworks, making it easier to handle exceptions consistently across your codebase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "67c2abc4-74e5-4f78-8310-aba942804463",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Exception'>\n",
      "    <class 'TypeError'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'ImportError'>\n",
      "        <class 'ModuleNotFoundError'>\n",
      "        <class 'zipimport.ZipImportError'>\n",
      "    <class 'OSError'>\n",
      "        <class 'ConnectionError'>\n",
      "            <class 'BrokenPipeError'>\n",
      "            <class 'ConnectionAbortedError'>\n",
      "            <class 'ConnectionRefusedError'>\n",
      "            <class 'ConnectionResetError'>\n",
      "                <class 'http.client.RemoteDisconnected'>\n",
      "        <class 'BlockingIOError'>\n",
      "        <class 'ChildProcessError'>\n",
      "        <class 'FileExistsError'>\n",
      "        <class 'FileNotFoundError'>\n",
      "        <class 'IsADirectoryError'>\n",
      "        <class 'NotADirectoryError'>\n",
      "        <class 'InterruptedError'>\n",
      "            <class 'zmq.error.InterruptedSystemCall'>\n",
      "        <class 'PermissionError'>\n",
      "        <class 'ProcessLookupError'>\n",
      "        <class 'TimeoutError'>\n",
      "        <class 'io.UnsupportedOperation'>\n",
      "        <class 'signal.itimer_error'>\n",
      "        <class 'socket.herror'>\n",
      "        <class 'socket.gaierror'>\n",
      "        <class 'ssl.SSLError'>\n",
      "            <class 'ssl.SSLCertVerificationError'>\n",
      "            <class 'ssl.SSLZeroReturnError'>\n",
      "            <class 'ssl.SSLWantWriteError'>\n",
      "            <class 'ssl.SSLWantReadError'>\n",
      "            <class 'ssl.SSLSyscallError'>\n",
      "            <class 'ssl.SSLEOFError'>\n",
      "        <class 'shutil.Error'>\n",
      "            <class 'shutil.SameFileError'>\n",
      "        <class 'shutil.SpecialFileError'>\n",
      "        <class 'shutil.ExecError'>\n",
      "        <class 'shutil.ReadError'>\n",
      "        <class 'urllib.error.URLError'>\n",
      "            <class 'urllib.error.HTTPError'>\n",
      "            <class 'urllib.error.ContentTooShortError'>\n",
      "        <class 'gzip.BadGzipFile'>\n",
      "    <class 'EOFError'>\n",
      "        <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'RuntimeError'>\n",
      "        <class 'RecursionError'>\n",
      "        <class 'NotImplementedError'>\n",
      "            <class 'zmq.error.ZMQVersionError'>\n",
      "            <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "        <class '_frozen_importlib._DeadlockError'>\n",
      "        <class 'threading.BrokenBarrierError'>\n",
      "        <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "            <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "        <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "        <class 'pkg_resources.ExtractionError'>\n",
      "        <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'NameError'>\n",
      "        <class 'UnboundLocalError'>\n",
      "    <class 'AttributeError'>\n",
      "        <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'SyntaxError'>\n",
      "        <class 'IndentationError'>\n",
      "            <class 'TabError'>\n",
      "    <class 'LookupError'>\n",
      "        <class 'IndexError'>\n",
      "        <class 'KeyError'>\n",
      "            <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "            <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "        <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ValueError'>\n",
      "        <class 'UnicodeError'>\n",
      "            <class 'UnicodeEncodeError'>\n",
      "            <class 'UnicodeDecodeError'>\n",
      "            <class 'UnicodeTranslateError'>\n",
      "        <class 'io.UnsupportedOperation'>\n",
      "        <class 'json.decoder.JSONDecodeError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'binascii.Error'>\n",
      "        <class '_hashlib.UnsupportedDigestmodError'>\n",
      "        <class 'calendar.IllegalMonthError'>\n",
      "        <class 'calendar.IllegalWeekdayError'>\n",
      "        <class 'dateutil.parser._parser.ParserError'>\n",
      "        <class 'pygments.util.ClassNotFound'>\n",
      "        <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "        <class 'email.errors.MessageDefect'>\n",
      "            <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "            <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "            <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "            <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "            <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "            <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "            <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "            <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "            <class 'email.errors.UndecodableBytesDefect'>\n",
      "            <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "            <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "            <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "            <class 'email.errors.HeaderDefect'>\n",
      "                <class 'email.errors.InvalidHeaderDefect'>\n",
      "                <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "                <class 'email.errors.NonPrintableDefect'>\n",
      "                <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "                <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "                <class 'email.errors.InvalidDateDefect'>\n",
      "        <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "        <class 'plistlib.InvalidFileException'>\n",
      "        <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "        <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "        <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "        <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "        <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "        <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "        <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "        <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "        <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "        <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "            <class 'pkg_resources.RequirementParseError'>\n",
      "        <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'ArithmeticError'>\n",
      "        <class 'FloatingPointError'>\n",
      "        <class 'OverflowError'>\n",
      "        <class 'ZeroDivisionError'>\n",
      "            <class 'decimal.DivisionByZero'>\n",
      "            <class 'decimal.DivisionUndefined'>\n",
      "        <class 'decimal.DecimalException'>\n",
      "            <class 'decimal.Clamped'>\n",
      "            <class 'decimal.Rounded'>\n",
      "                <class 'decimal.Underflow'>\n",
      "                <class 'decimal.Overflow'>\n",
      "            <class 'decimal.Inexact'>\n",
      "                <class 'decimal.Underflow'>\n",
      "                <class 'decimal.Overflow'>\n",
      "            <class 'decimal.Subnormal'>\n",
      "                <class 'decimal.Underflow'>\n",
      "            <class 'decimal.DivisionByZero'>\n",
      "            <class 'decimal.FloatOperation'>\n",
      "            <class 'decimal.InvalidOperation'>\n",
      "                <class 'decimal.ConversionSyntax'>\n",
      "                <class 'decimal.DivisionImpossible'>\n",
      "                <class 'decimal.DivisionUndefined'>\n",
      "                <class 'decimal.InvalidContext'>\n",
      "    <class 'SystemError'>\n",
      "        <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'Warning'>\n",
      "        <class 'UserWarning'>\n",
      "            <class 'getpass.GetPassWarning'>\n",
      "            <class 'IPython.core.formatters.FormatterWarning'>\n",
      "        <class 'EncodingWarning'>\n",
      "        <class 'DeprecationWarning'>\n",
      "            <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "        <class 'PendingDeprecationWarning'>\n",
      "        <class 'SyntaxWarning'>\n",
      "        <class 'RuntimeWarning'>\n",
      "            <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "            <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "            <class 'pkg_resources.PEP440Warning'>\n",
      "        <class 'FutureWarning'>\n",
      "            <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "        <class 'ImportWarning'>\n",
      "        <class 'UnicodeWarning'>\n",
      "        <class 'BytesWarning'>\n",
      "        <class 'ResourceWarning'>\n",
      "        <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "        <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'sre_parse.Verbose'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "        <class 'subprocess.CalledProcessError'>\n",
      "        <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "        <class 'concurrent.futures._base.CancelledError'>\n",
      "        <class 'concurrent.futures._base.TimeoutError'>\n",
      "        <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.TimeoutError'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "        <class 'zmq.error.ZMQError'>\n",
      "            <class 'zmq.error.ContextTerminated'>\n",
      "            <class 'zmq.error.Again'>\n",
      "            <class 'zmq.error.InterruptedSystemCall'>\n",
      "        <class 'zmq.error.ZMQBindError'>\n",
      "        <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "        <class '_pickle.PicklingError'>\n",
      "        <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "        <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "            <class 'traitlets.config.loader.ArgumentError'>\n",
      "        <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "        <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'zlib.error'>\n",
      "    <class 'tornado.util.TimeoutError'>\n",
      "    <class '_curses.error'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'termios.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class 'configparser.Error'>\n",
      "        <class 'configparser.NoSectionError'>\n",
      "        <class 'configparser.DuplicateSectionError'>\n",
      "        <class 'configparser.DuplicateOptionError'>\n",
      "        <class 'configparser.NoOptionError'>\n",
      "        <class 'configparser.InterpolationError'>\n",
      "            <class 'configparser.InterpolationMissingOptionError'>\n",
      "            <class 'configparser.InterpolationSyntaxError'>\n",
      "            <class 'configparser.InterpolationDepthError'>\n",
      "        <class 'configparser.ParsingError'>\n",
      "            <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'pexpect.exceptions.ExceptionPexpect'>\n",
      "        <class 'pexpect.exceptions.EOF'>\n",
      "        <class 'pexpect.exceptions.TIMEOUT'>\n",
      "    <class 'ptyprocess.util.PtyProcessError'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "        <class 'IPython.core.error.TryNext'>\n",
      "        <class 'IPython.core.error.UsageError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "        <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "        <class 'sqlite3.InterfaceError'>\n",
      "        <class 'sqlite3.DatabaseError'>\n",
      "            <class 'sqlite3.InternalError'>\n",
      "            <class 'sqlite3.OperationalError'>\n",
      "            <class 'sqlite3.ProgrammingError'>\n",
      "            <class 'sqlite3.IntegrityError'>\n",
      "            <class 'sqlite3.DataError'>\n",
      "            <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "        <class 'xml.dom.IndexSizeErr'>\n",
      "        <class 'xml.dom.DomstringSizeErr'>\n",
      "        <class 'xml.dom.HierarchyRequestErr'>\n",
      "        <class 'xml.dom.WrongDocumentErr'>\n",
      "        <class 'xml.dom.InvalidCharacterErr'>\n",
      "        <class 'xml.dom.NoDataAllowedErr'>\n",
      "        <class 'xml.dom.NoModificationAllowedErr'>\n",
      "        <class 'xml.dom.NotFoundErr'>\n",
      "        <class 'xml.dom.NotSupportedErr'>\n",
      "        <class 'xml.dom.InuseAttributeErr'>\n",
      "        <class 'xml.dom.InvalidStateErr'>\n",
      "        <class 'xml.dom.SyntaxErr'>\n",
      "        <class 'xml.dom.InvalidModificationErr'>\n",
      "        <class 'xml.dom.NamespaceErr'>\n",
      "        <class 'xml.dom.InvalidAccessErr'>\n",
      "        <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "        <class 'jedi.api.exceptions.InternalError'>\n",
      "        <class 'jedi.api.exceptions.WrongVersion'>\n",
      "        <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "        <class 'email.errors.MessageParseError'>\n",
      "            <class 'email.errors.HeaderParseError'>\n",
      "            <class 'email.errors.BoundaryError'>\n",
      "        <class 'email.errors.MultipartConversionError'>\n",
      "        <class 'email.errors.CharsetError'>\n",
      "    <class 'uu.Error'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "        <class 'http.client.NotConnected'>\n",
      "        <class 'http.client.InvalidURL'>\n",
      "        <class 'http.client.UnknownProtocol'>\n",
      "        <class 'http.client.UnknownTransferEncoding'>\n",
      "        <class 'http.client.UnimplementedFileMode'>\n",
      "        <class 'http.client.IncompleteRead'>\n",
      "        <class 'http.client.ImproperConnectionState'>\n",
      "            <class 'http.client.CannotSendRequest'>\n",
      "            <class 'http.client.CannotSendHeader'>\n",
      "            <class 'http.client.ResponseNotReady'>\n",
      "        <class 'http.client.BadStatusLine'>\n",
      "            <class 'http.client.RemoteDisconnected'>\n",
      "        <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "        <class 'psutil.NoSuchProcess'>\n",
      "            <class 'psutil.ZombieProcess'>\n",
      "        <class 'psutil.AccessDenied'>\n",
      "        <class 'psutil.TimeoutExpired'>\n",
      "    <class 'psutil._pslinux._Ipv6UnsupportedError'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "        <class 'xmlrpc.client.ProtocolError'>\n",
      "        <class 'xmlrpc.client.ResponseError'>\n",
      "        <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "            <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "        <class 'pkg_resources.VersionConflict'>\n",
      "            <class 'pkg_resources.ContextualVersionConflict'>\n",
      "        <class 'pkg_resources.DistributionNotFound'>\n",
      "        <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n"
     ]
    }
   ],
   "source": [
    "#Q2. Write a python program to print Python Exception Hierarchy.\n",
    "\n",
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(' ' * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "        \n",
    "print_exception_hierarchy(Exception)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5f2ccf88-6b4e-4ba6-9c42-5b23bdb27b81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Division by zero!\n"
     ]
    }
   ],
   "source": [
    "#Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "\n",
    "# 1)ZeroDivisionError: This exception is raised when attempting to divide a number by zero.\n",
    "\n",
    "try:\n",
    "    result = 10 / 0\n",
    "except ZeroDivisionError:\n",
    "    print(\"Error: Division by zero!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5cdb9d48-3152-4003-94a7-7ea32021e038",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2) OverflowError: This exception is raised when an arithmetic operation exceeds the maximum limit of a numeric type.\n",
    "\n",
    "import sys\n",
    "\n",
    "try:\n",
    "    result = sys.maxsize + 1\n",
    "except OverflowError:\n",
    "    print(\"Error: Numeric overflow!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09b94232-eb99-4370-a3b0-92720d2785b0",
   "metadata": {},
   "source": [
    " Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    " \n",
    "The LookupError class in Python is a base class for exceptions that occur when a lookup or indexing operation fails. It serves as the superclass for more specific lookup-related exception classes. Two common exceptions derived from the LookupError class are KeyError and IndexError. Let's explain these exceptions with examples:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8216ca7-cd0f-489a-9eac-115b883a8090",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Key not found!\n"
     ]
    }
   ],
   "source": [
    "# example in KeyError\n",
    "\n",
    "my_dict = {'a': 1, 'b': 2, 'c': 3}\n",
    "\n",
    "try:\n",
    "    value = my_dict['d']\n",
    "except KeyError:\n",
    "    print(\"Error: Key not found!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c422bbd-a2be-4151-9424-8603fe7549f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Invalid index!\n"
     ]
    }
   ],
   "source": [
    "# example in IndexError\n",
    "\n",
    "my_list = [1, 2, 3]\n",
    "\n",
    "try:\n",
    "    value = my_list[3]\n",
    "except IndexError:\n",
    "    print(\"Error: Invalid index!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1754fa75-50b7-4795-837f-f451b57db763",
   "metadata": {},
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "ImportError is an exception that is raised when there is an error importing a module or a specific attribute from a module. It is a subclass of the ImportError class. The ImportError class is a base class for all import-related exceptions.\n",
    "\n",
    "ModuleNotFoundError is a subclass of ImportError that specifically indicates that a module cannot be found. It was introduced in Python 3.6 to provide more precise error reporting for module imports.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c8ceea7-284a-4070-9523-5807c7759d2f",
   "metadata": {},
   "source": [
    "Q6. List down some best practices for exception handling in python.\n",
    "\n",
    "1) Catch Specific Exceptions: Catch specific exceptions rather than using a generic except block. This allows you to handle different types of exceptions differently and provide appropriate error messages or actions based on the specific exception.\n",
    "\n",
    "2) Use Multiple Except Blocks: Use multiple except blocks to handle different exceptions individually. This allows you to provide tailored error handling logic for each type of exception.\n",
    "\n",
    "3) Avoid Catching Exception as a Whole: Avoid catching the base Exception class without specifying the specific exception types. Catching the base Exception can make it difficult to diagnose and handle specific exceptions properly. Only catch the exceptions that you expect and know how to handle.\n",
    "\n",
    "Handle Exceptions Locally: Handle exceptions at the appropriate level of the code, close to where the exception occurs. This helps in keeping the code clean and allows for more focused error handling.\n",
    "\n",
    "Use Finally Block: Use the finally block to ensure that certain cleanup or finalization tasks are performed, regardless of whether an exception is raised or not. The code within the finally block will always execute, even if an exception occurs.\n",
    "\n",
    "Avoid Silencing Exceptions: Avoid empty except blocks that silently ignore exceptions. If an exception occurs and is not properly handled or logged, it can make debugging and troubleshooting more difficult. At the very least, log the exception or provide some form of error handling mechanism.\n",
    "\n",
    "Reraise Exceptions Judiciously: If you catch an exception but cannot handle it adequately at that level, consider reraising the exception using the raise statement. This allows the exception to propagate up the call stack to be caught and handled at a higher level where appropriate.\n",
    "\n",
    "Logging and Error Reporting: Use a logging framework to log exceptions and error messages. This helps in diagnosing issues during development and provides valuable information for troubleshooting in production environments. Additionally, consider integrating error reporting mechanisms to capture and analyze exceptions in production systems.\n",
    "\n",
    "Follow Python Naming Conventions: Follow Python naming conventions when naming custom exception classes. This helps in making the code more readable and understandable by adhering to standard naming conventions.\n",
    "\n",
    "Keep Exception Handling Simple: Keep exception handling logic concise and focused. Avoid complex nested try-except structures whenever possible. Simplify the logic by breaking it into smaller, manageable functions or using helper functions to handle specific exceptions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
